spec/models/bill_spec.rb

```diff
@@ -0,0 +1,174 @@
+require "rails_helper"
+
+RSpec.describe Bill, type: :model do
+  describe "validations" do
+    context "when valid attributes are provide" do
+      it "is valid with correct phone_number format" do
+        should allow_value("0901234567").for(:phone_number)
+      end
+
+      it "cancellation reason can be nil" do
+        should_not validate_presence_of(:cancellation_reason)
+      end
+
+      it "is valid with a total value" do
+        should validate_presence_of(:total)
+      end
+
+      it "is valid with total_after_discount" do
+        should validate_presence_of(:total_after_discount)
+      end
+    end
+
+    context 'when invalid attributes are provided' do
+      it "is invalid if phone_number is missing" do
+        should_not allow_value(nil).for(:phone_number)
+      end
+
+      it "is invalid with incorrect phone_number format" do
+        should_not allow_value("12345").for(:phone_number)
+      end
+
+      it "is invalid without a total value" do
+        should validate_presence_of(:total)
+      end
+
+      it "is invalid without total_after_discount" do
+        should validate_presence_of(:total_after_discount)
+      end
+    end
+  end
+
+  describe "enums" do
+    context "when enums are defined" do
+      it "defines status enum correctly" do
+        should define_enum_for(:status).with_values(wait_for_pay: 0, wait_for_prepare: 1, wait_for_delivery: 2, completed: 3, cancelled: 4)
+      end
+
+      it "defines cancellation_reason enum correctly" do
+        should define_enum_for(:cancellation_reason).with_values(out_of_stock: 0, wrong_quantity: 1, wrong_product: 2)
+      end
+    end
+
+    context "when enums are not defined" do
+      it "raises an error if status enum is not defined" do
+        should_not define_enum_for(:status).with_values(pending: 5, refunded: 6)
+      end
+
+      it "raises an error if cancellation_reason enum is not defined" do
+        should_not define_enum_for(:cancellation_reason).with_values(dont_like: 3, sad: 4)
+      end
+    end
+  end
+
+  describe "associations" do
+    context "when associations are correct" do
+      it "belongs to user" do
+        should belong_to(:user)
+      end
+
+      it "belongs to voucher optionally" do
+        should belong_to(:voucher).optional
+      end
+
+      it "has many bill_details" do
+        should have_many(:bill_details).dependent(:destroy)
+      end
+
+      it "has many products through bill_details" do
+        should have_many(:products).through(:bill_details)
+      end
+
+      it "has one address" do
+        should have_one(:address).dependent(:destroy)
+      end
+    end
+  end
+
+  describe "scopes" do
+    let!(:bill_1) { FactoryBot.create(:bill, status: :completed, created_at: Time.zone.now - 1.day) }

```

Comment:
let vs let! khác nhau như nào em

Reply:
dạ let thì nó sẽ được khởi tạo khi mình sử dụng nó, trong khi đó let! thì nó sẽ khởi tạo giá trị trước khi chạy quá trình test. Cụ thể hơn là khi mình khởi tạo let để test, một khối lệnh trong it nếu gọi biến đó 2 lần thì nó sẽ khởi tạo lại 2 lần, nếu biến đó có liên quan tới thời gian hiện tại thì khi khởi tạo 2 lần thì sẽ trả về 2 đối tượng giá trị khác nhau. Vì vậy em chọn let! để tạo vùng nhớ sẵn cho đối tượng.